### PR TITLE
implement `setproperty!` for modules

### PR DESCRIPTION
**Pull request by simeonschaub**
This replaces #44137. As discussed on triage, instead of supporting
modules in `setfield!`, this adds two new builtins `getglobal` and
`setglobal!` explicitly for reading and modifying module bindings. We
should probably consider `getfield(::Module, ::Symbol)` to be
soft-deprecated, but I don't think we want to add any warnings since
that will likely just annoy people.

### CODE REVIEW

**Comment by simeonschaub** [comment]
I wonder whether we should also export `getglobal` and `setglobal!`?

**Comment by vtjnash** [comment]
Is this done? What remains?

We export the others (fieldtype, getfield, setfield!, swapfield!, modifyfield!, replacefield!, etc) so I think it would be most consistent to export this too

**Comment by simeonschaub** [comment]
Yes, this just needs someone to review, other than that it's done on my part.

**Review by vtjnash — APPROVED** [review]
seems nearly complete to me

File: src/builtins.c [hunk 1]
```diff
@@ -1175,6 +1171,84 @@ JL_CALLABLE(jl_f_isdefined)
...
+    JL_NARGS(getglobal, 2, 3);
+    if (nargs == 3) {
+        JL_TYPECHK(getglobal, symbol, args[2]);
+        order = jl_get_atomic_order_checked((jl_sym_t*)args[2], 1, 0);
+    }
+    JL_TYPECHK(getglobal, module, args[0]);
+    JL_TYPECHK(getglobal, symbol, args[1]);
+    jl_value_t *v = jl_eval_global_var((jl_module_t*)args[0], (jl_sym_t*)args[1]);
+    if (order == jl_memory_order_notatomic)
+        jl_atomic_error("getglobal: module binding cannot be read non-atomically");
```
Comment (simeonschaub) [inline src/builtins.c:1188]:
```suggestion
    if (order == jl_memory_order_notatomic)
        jl_atomic_error("getglobal: module binding cannot be read non-atomically");
    jl_value_t *v = jl_eval_global_var((jl_module_t*)args[0], (jl_sym_t*)args[1]);
```

**Comment by N5N3** [comment]
Looks like this PR breaks the effect inference on `getfield`. MWE:
```julia
g(bc) = bc.re
@descend g(1+1im)
[(+c,+e,+n,+t)]g(bc) in Main at REPL[2]:1
│ ─ %-1  = invoke g(::Complex{Int64})::Int64
1 1 ─ %1 = Base.getfield(bc, :re)::Int64                                                                                           │╻ getproperty   
  └──      return %1                                                                                                               │ 
```
After this PR:
```julia
[(+c,!e,+n,+t)]g(bc) in Main at REPL[2]:1
│ ─ %-1  = invoke g(::Complex{Int64})::Int64
1 1 ─ %1 = Base.getfield(bc, :re)::Int64                                                                                           │╻ getproperty   
  └──      return %1
```

**Pull request merged**